src/hotspot/share/opto/loopnode.hpp

```diff
@@ -1266,26 +1270,92 @@ class PhaseIdealLoop : public PhaseTransform {
   // Per-Node transform
   virtual Node* transform(Node* n) { return nullptr; }
 
-  Node* loop_exit_control(Node* x, IdealLoopTree* loop);
-  Node* loop_exit_test(Node* back_control, IdealLoopTree* loop, Node*& incr, Node*& limit, BoolTest::mask& bt, float& cl_prob);
-  Node* loop_iv_incr(Node* incr, Node* x, IdealLoopTree* loop, Node*& phi_incr);
-  Node* loop_iv_stride(Node* incr, IdealLoopTree* loop, Node*& xphi);
-  PhiNode* loop_iv_phi(Node* xphi, Node* phi_incr, Node* x, IdealLoopTree* loop);
+  Node* loop_exit_control(const Node* x, const IdealLoopTree* loop) const;
+
+  struct LoopExitTest {
+    Node* cmp = nullptr;
+    Node* incr = nullptr;
+    Node* limit = nullptr;
+    const BoolTest::mask mask = BoolTest::illegal;
+    const float cl_prob = 0.0f;
+  };
+  LoopExitTest loop_exit_test(const Node* back_control, const IdealLoopTree* loop);
+
+  struct LoopIVIncr {
+    Node* incr = nullptr;
+    Node* phi_incr = nullptr;
+  };
+  LoopIVIncr loop_iv_incr(Node* old_incr, const Node* x, const IdealLoopTree* loop);
+
+  struct LoopIvStride {
+    Node* stride = nullptr;
+    Node* xphi = nullptr;
+  };
+  static LoopIvStride loop_iv_stride(const Node* incr);
 
-  bool is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_bt);
+  static PhiNode* loop_iv_phi(const Node* xphi, const Node* phi_incr, const Node* x);
+
+  bool try_convert_to_counted_loop(Node* head, IdealLoopTree*&loop, const BasicType iv_bt);
 
   Node* loop_nest_replace_iv(Node* iv_to_replace, Node* inner_iv, Node* outer_phi, Node* inner_head, BasicType bt);
   bool create_loop_nest(IdealLoopTree* loop, Node_List &old_new);
-#ifdef ASSERT
-  bool convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* loop);
-#endif
+
   void add_parse_predicate(Deoptimization::DeoptReason reason, Node* inner_head, IdealLoopTree* loop, SafePointNode* sfpt);
-  SafePointNode* find_safepoint(Node* back_control, Node* x, IdealLoopTree* loop);
+  SafePointNode* find_safepoint(Node* back_control, const Node* x, const IdealLoopTree* loop);
   IdealLoopTree* insert_outer_loop(IdealLoopTree* loop, LoopNode* outer_l, Node* outer_ift);
-  IdealLoopTree* create_outer_strip_mined_loop(BoolNode *test, Node *cmp, Node *init_control,
+  IdealLoopTree* create_outer_strip_mined_loop(Node *init_control,
                                                IdealLoopTree* loop, float cl_prob, float le_fcnt,
                                                Node*& entry_control, Node*& iffalse);
 
+  class CountedLoopConverter {
```

Comment:
Probably quite a subjective matter, but what about just naming it `CountedLoop`? Then you have `counted_loop.convert()`.

Reply:
What I had in mind was something like that (I'm also fine with `CountedLoopConverter`:
```
CountedLoop counted_loop(...);
counted_loop.convert();
return counted_loop.is_valid();
```
But maybe you can explain in more detail what the follow-up work will be and how you use this class again later?

Reply:
About the placement of this class. You could probably also move it out of the already huge `PhaseIdealLoop` class to make it a non-inner class, if there is not some strong coupling that you absolutely need.

Reply:
> [...] more detail what the follow-up work [...]

I was referring to [JDK-8336759](https://bugs.openjdk.org/browse/JDK-8336759) and #22449. Essentially I'm trying to achieve:
```cpp
CountedLoopConverter c(...);
if (!c.is_counted_loop() && iv_bt == T_INT and && limit_t == T_LONG) {
    // transform `i < (long) limit` to `i < (int) (long) limit` if limit < INT_MAX
}

if (c.is_counted_loop()) {
    c.convert();
}
```

Which I personally think makes more sense to confirm a counted loops before trying out the conversion.

> [...] move it out of the already huge `PhaseIdealLoop` [...]

Yes, I agree! I found myself questioning this nested class, too.